GVFS/GVFS.UnitTests.Windows/Windows/Upgrader/UpgradeOrchestratorTests.cs

```diff
@@ -78,7 +78,7 @@ public void GVFSDownloadError()
             this.ConfigureRunAndVerify(
                 configure: () =>
                 {
-                    this.Upgrader.SetFailOnAction(MockProductUpgrader.ActionType.GVFSDownload);
+                    this.Upgrader.SetFailOnAction(MockGitHubUpgrader.ActionType.GVFSDownload);

```

Comment:
Will create a separate PR for UT. This is what I have in mind -

**UpgradeOrchestrator & UpgradeVerb Tests**. Couple changes here.
1. What these classes test is high level upgrade flow, progress and error messaging. I don’t expect much changes in these tests. Currently the tests are a little polluted with details of Upgrade rings (which is specific to GitHubUpgrader) that probably will have to be moved out. 
2. The tests today are hard coded to use GitHubUpgrader instance only. New code has to be added so they could test the flow and messaging using IProductUpgrader instances. (The hard coding has to be removed and new code to test the flows using both GitHubUpgrader and NugetUpgrader instances has to be added.)

**ProductUpgrader Tests** - These are specific to GitHubUpgrader. The class has to be renamed. Probably consolidate test code spread across ProductUpgraderTests.cs and UpgradeTests.cs into one GitHubUpgraderTests.cs.

**New NugetUpgrader Tests** -  New tests to test specifics of Nuget upgrader.

Reply:
> Will create a separate PR for UT.

I was discussing this with @jamill a bit offline, but (at least for me) the state of the unit tests makes it harder to evaluate the interactions between the different components (e.g. the orchestrator, IProductUpgrader, etc.).

My understanding is that this PR's primary goal is to define the interaction between the components at a high level (without focusing on a specific `IProductUpgrader` implementation).  When I first saw `IProductUpgrader` I anticipated that there would be a pure “MockProductUpgrader” to test the interaction between the components.  As you pointed out though the product upgrader tests (and the `MockProductUpgrader`) are specific to `GitHubUpgrader`.  

For me, seeing all of the `GitHubUpgrader` specific code in the unit tests can make it tough to understand the high-level relationship between the components because the tests appear to be using GitHub specific concepts.  

For example, this test in `UpgradeVerbTests`:

```
[TestCase]
public void UpgradeAvailabilityReporting()
{
    this.ConfigureRunAndVerify(
        configure: () =>
        {
            this.SetUpgradeRing("Slow");
            this.Upgrader.PretendNewReleaseAvailableAtRemote(
                upgradeVersion: NewerThanLocalVersion,
                remoteRing: GitHubUpgrader.GitHubUpgraderConfig.RingType.Slow);
```

Uses GitHub specific upgrade rings, and reading it I'm not sure how the `UpgradeVerb` would work with non-GitHub `IProductUpgrader`.

> True. This is a cleanup that needs to be done, still there in my TODO list. It is not related to Nuget upgrader & this PR though

If there's still a TODO list for this feature to be production ready, should development be moved to a feature branch (with PRs into that feature branch) until the TODO list is complete?  

A feature branch would let you get some early feedback as you work through the different stages of this feature, and I would expect you'd be able to iterate more quickly (e.g. on `master` we should be making the required UT changes as we update the code, but they can be separate on a feature branch).

Reply:
renamed it to `UpgradeOrchestratorWithGitHubUpgraderTests.cs`